pkgs/native_assets_builder/lib/src/build_runner/build_runner.dart

```diff
@@ -689,50 +697,78 @@ ${e.message}
     );
     final dependenciesHashes = DependenciesHashFile(file: dependenciesHashFile);
     final lastModifiedCutoffTime = DateTime.now();
-    final bool mustCompile;
-    if (!await dependenciesHashFile.exists()) {
+    var mustCompile = false;
+    if (!await dependenciesHashFile.exists() || !await dartPathFile.exists()) {
       mustCompile = true;
     } else {
-      mustCompile =
-          (await dependenciesHashes.findOutdatedFileSystemEntity()) != null;
-    }
-    final bool success;
-    if (!mustCompile) {
-      success = true;
-    } else {
-      success = await _compileHookForPackage(
-        packageName,
-        scriptUri,
-        packageConfigUri,
-        workingDirectory,
-        includeParentEnvironment,
-        kernelFile,
-        depFile,
-      );
+      final previousDartExecutable =
+          Uri.file(await dartPathFile.readAsString());
+      if (previousDartExecutable != dartExecutable) {
+        mustCompile = true;
+        logger.info(
+          'Recompiling ${scriptUri.toFilePath()}, Dart executable changed.',
+        );
+      }
 
-      if (success) {
-        // Format: `path/to/my.dill: path/to/my.dart, path/to/more.dart`
-        final depFileContents = await depFile.readAsString();
-        final dartSources = depFileContents
-            .trim()
-            .split(' ')
-            .skip(1) // '<kernel file>:'
-            .map(Uri.file)
-            .toList();
-        final modifiedDuringBuild = await dependenciesHashes.hashFiles(
-          dartSources,
-          validBeforeLastModified: lastModifiedCutoffTime,
+      final outdatedFile =
+          await dependenciesHashes.findOutdatedFileSystemEntity();
+      if (outdatedFile != null) {
+        mustCompile = true;
+        logger.info(
+          'Recompiling ${scriptUri.toFilePath()}, '
+          '${outdatedFile.toFilePath()} changed.',
         );
-        if (modifiedDuringBuild != null) {
-          logger.severe('File modified during build. Build must be rerun.');
-        }
-      } else {
-        await dependenciesHashFile.delete();
       }
     }
+
+    if (!mustCompile) {
+      return (true, kernelFile, dependenciesHashFile);
+    }
+
+    final success = await _compileHookForPackage(
+      packageName,
+      scriptUri,
+      packageConfigUri,
+      workingDirectory,
+      includeParentEnvironment,
+      kernelFile,
+      depFile,
+    );
+    if (!success) {
+      await dependenciesHashFile.delete();
+      return (success, kernelFile, dependenciesHashFile);
+    }
+
+    final dartSources = await _readDepFile(depFile);
+    final modifiedDuringBuild =
+        await dependenciesHashes.hashFilesAndDirectories(
+      [
+        ...dartSources,
+        // If the Dart executable is replaced in-place, recompile.
+        dartExecutable,
```

Comment:
That's a great idea! Thanks @mkustermann!

(Also, if you put `out/ReleaseX64/` on the path instead of the bin folder, there won't be a `version` file at `../version`. But this is fine, it will just be hashed as `_hashNotExists`.)